apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1146 @@
+the top variants can be filtered out, and the lower variants will be used instead.
+If you don't have a compatible CUDA runtime at all, the fallback CPU wheel
+is used. So far, so good.
+
+However, consider a system with an older package manager version that does not
+support variants. Independently of the presence of CUDA runtime, all variant
+wheels are ignored there, and the CPU-only fallback is installed. However,
+this is not the most optimal solution. Prior to introducing variants,
+the published wheels may have featured both CUDA and CPU support (even if for a single
+CUDA version).
+
+This is where the null variant comes in. Consider the following instead.
+
+<figure>
+
+```rust
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu129.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu128.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu126.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-00000000.whl  // CPU-only
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64.whl           // CUDA 12.6 + CPU
+```
+
+<figcaption>Listing 5. Example wheel variant filenames, with a null variant and a fallback regular wheel, in order of preference</figcaption>
+</figure>
+
+What we added here is a null variant, with label `00000000`. Since it has
+no properties, it is always supported; that is, as long as variants
+are supported in the first place. This enables us to provide two different fallbacks:
+variant-enabled installers with no compatible CUDA runtime will use the CPU-only null
+variant, whereas installers without variant support (and therefore unable
+to determine the CUDA runtime version) will instead pick up the regular wheel, with both
+CUDA 12.6 and CPU support.
+
+Does it really matter? Well, that depends on how you look at it. Prior to wheel
+variant support being widely deployed, backwards compatibility is going to be important.
+And providing a different fallback may mean the difference between installing
+a 175 MiB CPU-only wheel, and a 850 MiB CUDA wheel, on a system
+without the CUDA runtime.
+
+## Multiple property values
+
+So far, we have discussed wheels where every feature has a single value declared.
+Say, a wheel that is optimized for a particular CPU architecture version,
+or has one CUDA version lower bound and one upper bound. The key point here
+is that for a variant to be compatible, all its features had to have a value
+compatible with the system in question. This logic is conjunctive, that is representing
+logical `AND`. You need to have a supported CPU, and a CUDA version that is
+not older than the lower bound, and a CUDA version (the same one) that is older
+than the upper bound.
+
+CPU instruction sets are an interesting case. Each
+instruction set is represented by a separate feature, with a single possible
+value of `on`. If the user's CPU supports a given instruction set, it supports
+said `on` value; if it doesn't, the feature has no compatible values. A wheel requiring
```

Comment:
This is perhaps too niche to go further into detail, but I'm left wondering if this will now by default create enormous list of cpu features in variant_features, or whether this is talking about a separate provider than the x64 v1-v4 aggregates

Reply:
Yeah, I think that's going too deeply into implementation details. Here the focus is on the possibility of defining properties like that. And well, technically speaking you could do either (or both), but our recommendation would be to do baseline + stuff above baseline.